File: C/queue.c
#include<stdio.h>
#include<malloc.h>

struct Queue {

    int front;
    int back;
    unsigned capacity;
    int numElements;
    int* array;
};

struct Queue* createQueue(unsigned capacity) {

    struct Queue* queue = malloc(sizeof(struct Queue));
    queue -> front = -1;
    queue -> back  = 0;
    queue -> numElements = 0;
    queue -> capacity = capacity;
    queue -> array = malloc(capacity * sizeof(int));
    return queue;
}

_Bool isEmpty(struct Queue* queue) {
    return queue -> numElements == 0;
}

_Bool isFull(struct Queue* queue) {
    return queue -> numElements == queue -> capacity;
}

void push(struct Queue* queue, int item) {

    if (isFull(queue)) {
        printf("Queue Overflow\n");
        return;
    }

    if (queue -> numElements == 0)
        queue -> front = -1;

    queue -> array[++queue -> front] = item;
    queue -> numElements++;
    return;
}

int pop(struct Queue* queue) {

    if (isEmpty(queue)) {
        printf("Queue Underflow\n");
        return -1;
    }

   queue -> numElements--;
   int element = queue -> array[queue -> back++];

   if (queue -> back == queue -> capacity)
       queue -> back = 0;

   return element;
}

int getSize(struct Queue* queue) {
    return queue -> numElements;
}


void show(struct Queue* queue) {

    int counter = 1;
    int index = queue -> back;

    while (counter <= queue -> numElements) {
        printf("%d\t", queue -> array[index]);
        counter++;
        index++;
    }

    printf("\n");
    return;
}



int main(int argc, char* argv[]) {

    struct Queue* queue = createQueue(3);
    push(queue, 1);
    push(queue, 2);
    push(queue, 3);
    show(queue);
    printf("%d\n", pop(queue));
    push(queue, 4);
    show(queue);
    printf("%d\n", pop(queue));
    printf("%d\n", pop(queue));
    printf("%d\n", pop(queue));
    return 0;

}

File: Python/queue.py
class Queue:

    def __init__(self, capacity: int):

        self.length = 0
        self.capacity = capacity
        self.head = 0
        self.tail = 0
        self.queue_arr = [0 for i in range(capacity)]

    def enqueue(self, item: int):

        if self.length == self.capacity:
            print("Queue Overflow")
            return

        if queue.tail == queue.capacity:
            queue.tail = 0

        self.queue_arr[self.tail] = item
        self.tail = self.tail + 1
        self.length = self.length + 1
        return

    def dequeue(self) -> int:

        if self.length == 0:
            print("Queue Underflow")
            return -1

        if self.head == self.capacity:
            self.head = 0

        item = self.queue_arr[self.head]
        self.head = self.head + 1
        self.length = self.length - 1

        return item

    def peek(self) -> int:
        
        if self.length == 0:
            return -1

        return self.queue_arr[self.head - 1]
